### PR TITLE
HIP-0000 Token Create through HTS Precompiled Contract

### PR DESCRIPTION
**Pull request by stoqnkpL**
Signed-off-by: Stoyan Panayotov <stoyan.panayotov@limechain.tech>

**Description**:

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: assets/hip-206/solidity/IHederaTokenService.sol [hunk 1]
```diff
@@ -49,6 +49,168 @@ interface IHederaTokenService {
...
+        // will belong to the given contract, since it could be running another contract's code via 
+        // <tt>delegatecall</tt>. So setting this key is a more permissive version of setting the
+        // contractID key, which also requires the code in the active message frame belong to the
+        // the contract with the given id.)
+        address delegatableContractId;
+    }
+
+    /// A convenience struct used to group different types of Keys we want to set to a Token during
+    /// token creation.
+    struct TokenKeys{
```
Comment (shemnon) [inline assets/hip-206/solidity/IHederaTokenService.sol:184]: Rather than a struct with all of the possible keys what about a struct with a key type (one of admin, kyc, ...) and then the Key.  The API would accept an array of this key struct.

And then, what if we made the type actually a bit field, so you could set multiple keys with one entry.  Odds are most contracts will set all to one, or one to each.  For the all to one scenario we save a lot of gas.

**Review by shemnon — COMMENTED** [review]
Before we "lock in" this API I think we should see what the solidity code that exercises these APIs would look like for common scenarios.  Some constructs may be very unweildy to build.

Two simple scenarios:
* Token with no custom fees, with a user account as admin, contract as supply and pause key.
* NFT token with a royalty fee, contract has the mint and admin key.

Also, do we need APIs to update fee schedule? And Pause a token, as well as freeze/kyc/wipe accounts?  Maybe APIs to do admin operations (like update keys I guess?)

File: assets/hip-206/solidity/IHederaTokenService.sol [hunk 2]
```diff
@@ -49,6 +49,148 @@ interface IHederaTokenService {
...
+        // will be created for the token.
+        // 0th bit: adminKey
+        // 1st bit: kycKey
+        // 2nd bit: freezeKey
+        // 3rd bit: wipeKey
+        // 4th bit: supplyKey
+        // 5th bit: feeScheduleKey
+        // 6th bit: pauseKey
+        // 7th bit: ignored
+        uint8 keyType;
```
Comment (shemnon) [inline assets/hip-206/solidity/IHederaTokenService.sol:174]: Unless this is a "packed" struct we should kale it at least uint16.  since it's not intended to "go over the wire" there is no need to pack it. perhaps just `uint`?

File: assets/hip-206/solidity/IHederaTokenService.sol [hunk 3]
```diff
@@ -133,6 +275,44 @@ interface IHederaTokenService {
...
+    /// @return bool whether the create was successful
+    /// @return address the address of the created token
+    function createFungibleToken(
+        HederaToken memory token, 
+        uint initialTotalSupply, 
+        uint decimals, 
+        Expiry memory expiry, 
+        FixedFee[] memory fixedFees, 
+        FractionalFee[] memory fractionalFees, 
+        Key[] memory keys) external returns (bool success, bytes tokenAddress);
```
Comment (shemnon) [inline assets/hip-206/solidity/IHederaTokenService.sol:296]: obsessive nit: I would put `keys` earlier in the list because they are more important.  Before the custom fees.  Possibly just before expiry.  Custom fees may be empty and I would want mandatory fields before optional.

File: assets/hip-206/solidity/TokenCreateContract.sol [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        IHederaTokenService.FractionalFee[] memory fractionalFees;
+
+        IHederaTokenService.HederaToken memory myToken;
+        myToken.name = "MyToken";
+        myToken.symbol = "MTK";
+        myToken.treasury = address(this);
+        myToken.tokenKeys = keys;
+
+        (bool success, address token, bytes memory errorMessage) = 
+            HederaTokenService.createFungibleToken(myToken, 200, 8, fixedFees, fractionalFees);
```
Comment (shemnon) [inline assets/hip-206/solidity/TokenCreateContract.sol:49]: I think we should have a varient without custom fees, or fold it into the token description.

File: assets/hip-206/solidity/TokenCreateContract.sol [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        // turn on bit corresponding to admin key type
+        adminKeyType = adminKeyType.setBit(0);
+        // set the value of the key to be inherited from the calling account
+        adminKeyValue.inheritAccountKey = true;
+        
+        // create TokenKey of types supplyKey and pauseKey with value a contract address passed as function arg
+        uint supplyPauseKeyType;
+        IHederaTokenService.KeyValue memory supplyPauseKeyValue;
+        // turn on bits corresponding to supply and pause key types
+        supplyPauseKeyType = supplyPauseKeyType.setBit(4);
```
Comment (shemnon) [inline assets/hip-206/solidity/TokenCreateContract.sol:28]: We should have constants for these in IHederaTokenService instead of relying on magic numbers.

File: assets/hip-206/solidity/TokenCreateContract.sol [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+        IHederaTokenService.HederaToken memory myToken;
+        myToken.name = "MyNFT";
+        myToken.symbol = "MNFT";
+        myToken.treasury = address(this);
+        myToken.tokenKeys = keys;
+
+        // create the token through HTS with default expiry and royalty fees;
+        (bool success, address token, bytes memory errorMessage) = 
+            HederaTokenService.createNonFungibleToken(myToken, fixedFees, fractionalFees, royaltyFees);
```
[the remark this answers is not in the document]
Comment (stoqnkpL) [reply assets/hip-206/solidity/TokenCreateContract.sol:93]: I thought about folding the custom fees into the `HederaToken` object, but I see two problems with that:

- `RoyaltyFee` contains a `FixedFee` within its structure and I'm not sure Solidity will handle having a struct containing both FixedFees and RoyaltyFees with FixedFees within them well
- As the `HederaToken` struct gets more and more complicated it becomes harder for decoding and testing all edge-cases

But I do agree that we need to simplify the interface and adding two more create functions with no custom fee arguments is necessary.

**Comment by stoqnkpL** [comment]
One question that came up internally today is whether we really want to allow updates of tokens for the threshold keys with this HIP or just limit the Token create HIP to the create functions with 1 of 1 keys and introduce the Update functionality as part of a separate HIP

**Comment by SimiHunjan** [comment]
Limit the Token create HIP to the create functions with 1 of 1 keys for this release

File: HIP/hip-358.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+| ---------- | -------------------------------------------------------------------------------------------------|--------------------------|
+|  | `createFungibleToken(HederaToken, uint, uint)`                                                             | `(bool, addess, bytes)`  |
+|  | `createFungibleTokenWithCustomFees(HederaToken, uint, uint, FixedFee[], FractionalFee[])`                  | `(bool, addess, bytes)`  |
+|  | `createNonFungibleToken(HederaToken, uint, uint)`                                                          | `(bool, addess, bytes)`  |
+|  | `createNonFungibleTokenWithCustomFees(HederaToken, uint, uint, FixedFee[], FractionalFee[], RoyaltyFee[])` | `(bool, addess, bytes)`  |
+
+### Precompile Gas Costs
+
+Gas will be priced so that it will be cheaper to do calls directly to HTS rather
+than through HSCS where such calls are possible.
```
Comment (shemnon) [inline HIP/hip-358.md:75]: We will need actual numbers before this can go final, but it won't block initial submission.

Can you add a table with the functiongroups and gas cost column with `TBD` as the price?

**Review by shemnon — DISMISSED** [review]

**Review by sergmetelin — CHANGES_REQUESTED** [review]

**Review by shemnon — APPROVED** [review]

File: assets/hip-206/solidity/IHederaTokenService.sol [hunk 8]
```diff
@@ -160,8 +160,12 @@ interface IHederaTokenService {
         uint32 amount;
 
         // The denomination of the fee; taken as hbar if left unset and taken as the id
-        // of the newly created token if set to the sentinel value of 0x0
+        // of the newly created token if set to the sentinel value of 0x0 (since there
+        // is no notion of unset or null value in Solidity and the default value of an
+        // address variable is 0x0, the `isTokenIdSet` variable is used to distinguish
+        // between the two cases)
         address tokenId;
+        bool isTokenIdSet;
```
Comment (shemnon) [inline assets/hip-206/solidity/IHederaTokenService.sol:168]: Why would tokenId ever be un-set?  If passed in as an argument I think it's reasonable to expect that the required fields are set.  Is there any meaning to a fixed fee with an amount, a fee collector, and an unset tokenI?

Comment (stoqnkpL) [reply assets/hip-206/solidity/IHederaTokenService.sol:168]: Yes, there are some special meanings in the protobufs values 0 and unset for this field:

- if tokenId is 0, then use the token that's being created in this tx's id
- if tokenId is unset, then use hbars
- if tokenId is set to non-zero number, then use the id specified

When decoding Solidity input args however it's impossible to distinguish between cases 1 and 2 since missing value and zero value are both decoded to 0x0.

I do think that we can make it easier to read and understand in the Solidity struct though, maybe something like:

address tokenId;
bool useHbarsForPayment;
bool useCurrentTokenForPayment;

and specify that exactly one of the fields must be populated. If more than one field is populated we would fail the decoding, set error message in the message frame and return null from the precompile invocation.

**Review by sergmetelin — APPROVED** [review]
LGTM

File: HIP/hip-358.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+| hash       | effective signature                                                                       | return           |
+| ---------- | ------------------------------------------------------------------------------------------|------------------|
+| `7812a04b` | `createFungibleToken(HederaToken, uint, uint)`                                            | `(int, addess)`  |
+| `4c381ae7` | `createFungibleTokenWithCustomFees(HederaToken, uint, uint, FixedFee[], FractionalFee[])` | `(int, addess)`  |
+| `9dc711e0` | `createNonFungibleToken(HederaToken)`                                                     | `(int, addess)`  |
+| `5bc7c0e6` | `createNonFungibleTokenWithCustomFees(HederaToken, FixedFee[], RoyaltyFee[])`             | `(int, addess)`  |
+
+### Precompile Gas Costs
+
+In order to allow creation of token through the precompiled contract, a new mechanism for payment is introduced - precompile calls are partially payed by Hbar value transfers.
```
Comment (shemnon) [inline HIP/hip-358.md:67]: We will have to update `com.hedera.services.store.contracts.HederaWorldState#TOKEN_CALL_REDIRECT_CONTRACT_BINARY` - make it payable.  The current evm code will fail when any value is passed in.  We will need to move the value check to another part of the HTS precompile.

File: HIP/hip-358.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+## Specification
+
+### EVM Precompile extension
+
+The [Solidity file for development](../assets/hip-206/solidity/IHederaTokenService.sol) is updated with new types and function signatures that the precompile will respond to. It is included in this HIP by reference.
+
+### Solidity Function Signatures 
+
+The following structs are defined to simplify the interface between solidity and the HTS precompile:
```
Comment (shemnon) [inline HIP/hip-358.md:44]: For the data structures we should describe what each parameter does, not just the type.  Probably in a markdown prose section not in the table.

File: HIP/hip-358.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+| `9dc711e0` | `createNonFungibleToken(HederaToken)`                                                     | `(int, addess)`  |
+| `5bc7c0e6` | `createNonFungibleTokenWithCustomFees(HederaToken, FixedFee[], RoyaltyFee[])`             | `(int, addess)`  |
+
+### Precompile Gas Costs
+
+In order to allow creation of token through the precompiled contract, a new mechanism for payment is introduced - precompile calls are partially payed by Hbar value transfers.
+There is a base cost of 100,000 Gas for each call to the create functions. Additional Hbar cost is required to be transferred as `value` to the precompile call. The Hbar price will be calculated using the following formula:
+
+The total cost of HTS Token create through the precompile is equal to the cost of a HAPI Create Token Transaction with similar arguments at that instance plus a 20% premium. 
+The Hbars required to be transferred as value in the precompile call will be equal to the total cost required minus the Hbar cost for the base 100,000 gas cost at that instance.
```
Comment (shemnon) [inline HIP/hip-358.md:71]: > 100,000 gas cost

"100,000 gas cost converted into hbar according to the current gas price"

**Review by shemnon — APPROVED** [review]
This has the core of the idea, but before it comes out of review we need to firm up the meaning of the field in the HIP and not via the supplementary solidity fiels.

**Review by SimiHunjan — APPROVED** [review]
